Enable shadowrealm testing for url api

rwaldron opened the pull request:
Starting with tests that don't have additional dependencies.

rwaldron committed:
Enable shadowrealm testing for url api

Starting with tests that don't have additional dependencies.

annevk reviewed (commented):
So the HTML PR for this is still outstanding, right? And there was also talk of defining some kind of IDL construct which I don't think has happened. What is the standardization status of this feature?

rwaldron commented:
@annevk thanks for the review! Changes made as requested

annevk commented:
So the problem from my perspective is that we typically land test changes alongside the specification PR change. Landing test changes ahead of the specification being formalized is not something we have precedent for, apart from when tests have a `.tentative` suffix.

rwaldron commented:
@annevk I understand, and it should be fine to leave the PR open until the spec change work is complete, if that's what you'd prefer. I will follow up here when the time is right.

ptomato committed:
Split URL tests into ShadowRealm-applicable and not-applicable files

Some tests required interfaces such as Request, Response, or FormData that
are not available in ShadowRealm. These should be split into separate test
files that are not run in ShadowRealm scopes.

ptomato committed:
Split tests for failure to parse base URL into their own file

In url/failure.html there were tests passing invalid URLs to the second
parameter of the URL constructor. Split these into their own file so they
can be run in workers and ShadowRealms, separately from the other tests in
url/failure.html that use the current document's URL.

ptomato committed:
Enable ShadowRealm testing for more URL APIs

ptomato commented:
Rebased and added some additional changes.

annevk commented on url/urlencoded-parser.any.js:
```diff
@@ -1,68 +1,19 @@
+  });
+}
+
+promise_test(() => fetch_json("resources/urlencoded-parser-data.json").then(runTests), "Loading data…");
```
This line is typically at the top. Let's keep it there?

annevk commented on url/urlencoded-parser-request-response.any.js:
```diff
@@ -0,0 +1,31 @@
+  });
+}
+
+promise_test(() => fetch_json("resources/urlencoded-parser-data.json").then(runTests), "Loading data…");
```
Let's keep these at the top?

annevk commented on url/url-setters.any.js:
```diff
@@ -6,7 +7,7 @@
 // Keep this file in sync with url-setters-a-area.window.js.
 
-promise_test(() => fetch("resources/setters_tests.json").then(res => res.json()).then(runURLSettersTests), "Loading data…");
+promise_test(() => fetch_json("resources/setters_tests.json").then(runURLSettersTests), "Loading data…");
```
Where does fetch_json come from?

ptomato replied:
I added it in [RFC 177](https://github.com/web-platform-tests/rfcs/blob/master/rfcs/fetch_json_et_al.md#add-fetch_json-test-harness-api) earlier this year.

annevk reviewed (commented):
Assuming you carefully reviewed the before and after results this seems okay. However, I'm not sure what the impact of this is on Interop 2024. This late in the year it might require someone else to double check this doesn't end up impacting the Interop scores in some weird way.

So not approving until we're fully clear on that.

annevk commented:
Thanks @ptomato.

I think for split tests we should only include the any.html and not the any.worker.html. That way we don't increase or decrease the overall number of tests in Interop. Assuming the labeling is done well that should then amount to this being a no-op, which is what we'd want at this point. Which would reduce your list down to:

* IdnaTestV2.any.html
* urlencoded-parser-request-response.any.html
* url-constructor-base-failure.any.html

And we'd continue to include what we include now.

@gsnedders @jgraham @foolip please let me know if you have any concerns.

annevk reviewed (approved):
Approving, but please give @jgraham @gsnedders @foolip one week to weigh in before merging this. And if this does turn out to not be a no-op for Interop 2024 it'll have to be reverted.

jgraham commented:
The scoring system is sum over all tests of fraction of subtests passing divided by the total number of tests e.g. if you have two tests with scores 2/2 and 10/20 then the Interop score will be (1 + 0.5)/2 = 0.75, not 12/22.

So I think to land this we also need a PR against wpt-metadata to add the labels to whatever new tests require them, so that it can land right afterwards (I think it will be blocked until this PR is landed due to the missing tests). And it needs to be clear that the interop scores are not going to change once both PRs are landed.

ptomato commented:
Ah, that does change things somewhat, as test failures in a file with few tests have a much larger impact than test failures in a file with many tests.

I've done a quick spreadsheet calculating the scores that way and given Anne's proposed list of wpt-metadata changes from https://github.com/web-platform-tests/wpt/pull/41985#issuecomment-2457344095. I'll spare you the details but here is the result.

|  | Chrome | Firefox | Safari |
| --- | --- | --- | --- |
| Before | 95.70% | 99.33% | 99.79% |
| After | 95.82% | 99.19% | 99.80% |
| Difference | +0.12% | −0.15% | +0.02% |

Not sure if these changes are small enough to be considered just noise, or this would be a problem?

An alternative would be to only replace `IdnaTestV2.window.html` with `IdnaTestV2.any.html` in wpt-metadata. That would give only +0.01% to Firefox and leave the other two scores unchanged to two decimal places. (i.e., hundredths of a percent.) It'd be less accurate IMO because it would actually move existing tests out from under the interop umbrella, but it'd be the most no-op possible given the constraints of the scoring system.

gsnedders reviewed (changes requested):
> Not sure if these changes are small enough to be considered just noise, or this would be a problem?

Given it's not a no-op change to Interop, this needs an [Interop Test Change Proposal](https://github.com/web-platform-tests/interop/issues/new?assignees=&labels=test-change-proposal&projects=&template=test-change-proposal.yml) issue opened before it (and the as-of-yet-unwritten wpt-metadata PR) can land together.

[This view](https://wpt.fyi/results/url?sha=f382c0ce93&max-count=2&product=chrome&product=firefox&product=safari&view=interop&q=label%3Ainterop-2023-url%20or%20IdnaTestV2.any.html%20or%20urlencoded-parser-request-response.any.html%20or%20url-constructor-base-failure.any.html%20is%3Adifferent) exaggerates the differences this PR makes (through the inclusion of `is:different`), but does show there are meaningful changes. ([This](https://wpt.fyi/results/url?sha=f382c0ce93&max-count=2&product=chrome&product=firefox&product=safari&view=interop&q=label%3Ainterop-2023-url%20or%20IdnaTestV2.any.html%20or%20urlencoded-parser-request-response.any.html%20or%20url-constructor-base-failure.any.html) shows the proposed Interop results, I believe.)

(FWIW: I'm not opposed to this change being made, it just needs to go through the right process given the Interop impact.)

ptomato commented:
Thanks for the pointers and for the nice link showing the results; much better than doing it in a spreadsheet.

wpt-metadata PR: https://github.com/web-platform-tests/wpt-metadata/pull/6930
Test change proposal: https://github.com/web-platform-tests/interop/issues/904

ptomato commented:
There haven't been any comments on the Interop test change proposal for a week, so I'm assuming folks are OK with it. What would be the next step in that process?
